diff --git a/assets/src/index.tsx b/assets/src/index.tsx
index 7b2972ea..d4c3e49e 100644
--- a/assets/src/index.tsx
+++ b/assets/src/index.tsx
@@ -9,7 +9,7 @@ import { PyGetRootColors, PyGetStartupConfig } from './utils/ffi';
 import { onWindowCreatedCallback, patchMissedDocuments } from './patcher';
 import { MillenniumSettings } from './settings';
 import { NotificationService } from './update-notification-service';
-import { EUIMode } from '@steambrew/client/build/globals/steam-client/shared';
+//import { EUIMode } from '@steambrew/client/build/globals/steam-client/shared';
 import { MillenniumDesktopSidebar } from './quick-access';
 import { DesktopMenuProvider } from './quick-access/DesktopMenuContext';
 
@@ -67,7 +67,7 @@ export default async function PluginMain() {
                 <MillenniumDesktopSidebar />
             </DesktopMenuProvider>
         ),
-        EUIMode.Desktop,
+        7,
     );
 
     // @ts-ignore
